src/core/lib/transport/promise_endpoint.h

```diff
@@ -0,0 +1,315 @@
+ */
+
+#ifndef GRPC_CORE_LIB_TRANSPORT_PROMISE_ENDPOINT_H
+#define GRPC_CORE_LIB_TRANSPORT_PROMISE_ENDPOINT_H
+
+#include <grpc/support/port_platform.h>
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include <functional>
+#include <memory>
+#include <utility>
+
+#include "absl/base/thread_annotations.h"
+#include "absl/status/status.h"
+#include "absl/status/statusor.h"
+#include "absl/types/optional.h"
+
+#include <grpc/event_engine/event_engine.h>
+#include <grpc/event_engine/slice.h>
+#include <grpc/event_engine/slice_buffer.h>
+#include <grpc/slice_buffer.h>
+#include <grpc/support/log.h>
+
+#include "src/core/lib/gprpp/sync.h"
+#include "src/core/lib/promise/activity.h"
+#include "src/core/lib/promise/poll.h"
+#include "src/core/lib/slice/slice.h"
+#include "src/core/lib/slice/slice_buffer.h"
+
+namespace grpc {
+
+namespace internal {
+
+class PromiseEndpoint {
+ public:
+  PromiseEndpoint(
+      std::unique_ptr<grpc_event_engine::experimental::EventEngine::Endpoint>
+          endpoint,
+      grpc_core::SliceBuffer already_received);
+  ~PromiseEndpoint();
+
+  auto Write(grpc_core::SliceBuffer data) {
+    grpc_core::MutexLock lock(&write_mutex_);
+
+    /// Previous write result has not been polled.
+    GPR_ASSERT(!write_result_.has_value());
+
+    /// TODO: Is there a better way to convert?
+    grpc_slice_buffer_swap(data.c_slice_buffer(),
+                           write_buffer_.c_slice_buffer());
+
+    const std::function<void(absl::Status)> write_callback =
+        [this](absl::Status status) {
+          grpc_core::MutexLock lock(&write_mutex_);
+          write_result_ = status;
+          write_waker_.Wakeup();
+        };
+    endpoint_->Write(write_callback, &write_buffer_,
+                     nullptr /* uses default arguments */);
+
+    return [this]() -> grpc_core::Poll<absl::Status> {
+      grpc_core::MutexLock lock(&write_mutex_);
+      if (!write_result_.has_value()) {
+        write_waker_ = grpc_core::Activity::current()->MakeNonOwningWaker();
+        return grpc_core::Pending();
+      } else {
+        const auto ret = *write_result_;
+        write_result_.reset();
+        return ret;
+      }
+    };
+  }
+
+  auto Read(size_t num_bytes) {
+    grpc_core::MutexLock lock(&read_mutex_);
+
+    /// Previous read result has not been polled.
+    GPR_ASSERT(!read_result_.has_value());
+
+    /// Should not have pending reads
+    GPR_ASSERT(pending_read_buffer_.Count() == 0u);
+
+    if (read_buffer_.Count() < num_bytes) {
+      const std::function<void(absl::Status)> read_callback =
+          [this, num_bytes, &read_callback](absl::Status status) {
+            if (!status.ok()) {
+              pending_read_buffer_.Clear();
+              temporary_read_buffer_.Clear();
+
+              grpc_core::MutexLock lock(&read_mutex_);
+              read_result_ = status;
+              read_waker_.Wakeup();
+            } else {
+              while (temporary_read_buffer_.Count() > 0u) {
```

Comment:
Instead of this loop, you can use the MoveFirstNBytesIntoSliceBuffer method defined in grpc_event_engine::experimental::SliceBuffer

Reply:
I think the reason for this is that I cannot use `grpc_event_engine::experimental::SliceBuffer`.

Reply:
You could use `grpc_event_engine::experimental::SliceBuffer` internally and convert at the end if that helps (and I think we'll be able to make that final conversion 'free' soon)